### PR TITLE
Desktop UI for Solana signAndSendTransaction, signTransaction, and signAllTransactions provider APIs

### PR DESCRIPTION
**Pull request by josheleonard**
Resolves <https://github.com/brave/brave-browser/issues/23369>
Resolves <https://github.com/brave/brave-browser/issues/23434>

Security review request: https://github.com/brave/security/issues/910

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. Create a new `Wallet`, select `Solana Devnet` and then create a `Solana Account` if needed.
2. Go to https://solfaucet.com/ , paste your `Solana` address and click `Devnet` to get some `Devnet Solana Tokens`
3. Go back to the `Brave Wallet` and make sure you have received your `Devnet` tokens in your wallet.
4. Ask @Douglashdaniel for a shared test wallet or message @yrliou to send you some `SPL` tokens on the `Solana Devnet` (you will need to add this token as a custom token in the `Edit Visible Asset` modal.)
5. Go to https://pwgoom.csb.app/ and click `Connect`, connect your wallet!

## Testing - Sign and Send Transaction

1. Click `Sign and Send Transaction`.

Screen Shot 2022-06-16 at 11 08 29 PM

2. Check and make sure `Transaction` info is displayed.

Screen Shot 2022-06-16 at 10 37 46 PM

3. Click `Details` and make sure transaction `Details` are displayed.

Screen Shot 2022-06-16 at 10 38 04 PM

4. Click `Confirm` and make sure you can see the `Transaction Details`

Screen Shot 2022-06-16 at 10 41 50 PM

## Testing - Sign and Send Transaction (Request)

1. Click on `Sign and Send Transaction (Request)`

Screen Shot 2022-06-16 at 11 08 40 PM

2. Follow the same instructions for `Testing - Sign and Send Transaction` from `2-4`

## Testing - Sign and Send SPL Token Transaction

1. Click `Sign and Send SPL Token Transaction`.

Screen Shot 2022-06-16 at 11 11 16 PM

2. Check and make sure `Transaction` info is displayed.

Screen Shot 2022-06-17 at 9 56 46 AM

3. Click `Details` and make sure transaction `Details` are displayed.

Screen Shot 2022-06-17 at 9 57 00 AM

4. Click `Confirm` and make sure you can see the `Transaction Details`

Screen Shot 2022-06-17 at 9 57 14 AM

## Testing - Sign Transaction

1. Click `Sign Transaction`.

Screen Shot 2022-06-16 at 11 30 05 PM

2. Review `Sign at your own risk` warning and click `Continue`

Screen Shot 2022-06-16 at 10 55 59 PM

3. Review `Details` and click `Sign`

Screen Shot 2022-06-16 at 10 56 18 PM

## Testing - Sign Transaction (Request)

1. Click on `Sign Transaction (Request)`

Screen Shot 2022-06-16 at 11 31 00 PM

2. Follow the same instructions for `Sign Transaction` from `2-3`

## Testing - Sign All Transaction (multiple)

1. Click on `Sign All Transaction (multiple)`

Screen Shot 2022-06-16 at 11 31 44 PM

2. Review `Sign at your own risk` warning and click `Continue`

Screen Shot 2022-06-16 at 10 55 59 PM

3. Review `Details` and make sure that there are `multiple` transactions, then click `Sign`

Screen Shot 2022-06-16 at 11 03 15 PM

## Testing - Sign All Transaction (multiple) (Request)

1. Click on `Sign All Transaction (multiple) (Request)`

Screen Shot 2022-06-16 at 11 32 17 PM

2. Follow the same instructions for `Sign All Transaction (multiple)` from `2-3`

## Testing - Sign All Transaction (single)

1. Click on `Sign All Transaction (single)`

Screen Shot 2022-06-16 at 11 32 46 PM

2. Review `Sign at your own risk` warning and click `Continue`

Screen Shot 2022-06-16 at 11 25 33 PM

3. Review `Details` and make sure that there is a `single` transaction, then click `Sign`

Screen Shot 2022-06-16 at 11 25 45 PM

## Testing - Sign All Transaction (single) (Request)

1. Click on `Sign All Transaction (single) (Request)`

Screen Shot 2022-06-16 at 11 33 28 PM

2. Follow the same instructions for `Sign All Transaction (single)` from `2-3`

### CODE REVIEW

File: components/brave_wallet_ui/utils/solana-instruction-utils.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  & Solana.CreateAccountParams
+  & Solana.CreateAccountWithSeedParams
+  & Solana.InitializeNonceParams
+  & Solana.TransferParams
+  & Solana.TransferWithSeedParams
+  & Solana.WithdrawNonceParams
+
+export type SolanaInstructionParamKeys = keyof SolanaInstructionParams
+
+export const getSolanaSystemInstructionParamsAndType = ({
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/utils/solana-instruction-utils.ts:49]: Should we have tests for these utils?

File: components/brave_wallet_ui/utils/solana-program-id-utils.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+// utils
+import { getLocale } from '$web-common/locale'
+
+/**
+ * see: https://docs.solana.com/developing/runtime-facilities/programs
+ * @param programId Solana.PublicKey
+ * @returns Name of program
+ */
+export const getSolanaProgramIdName = (programId: Solana.PublicKey) => {
```
Comment (Douglashdaniel) [inline components/brave_wallet_ui/utils/solana-program-id-utils.ts:16]: Ditto!

**Review by Douglashdaniel — COMMENTED** [review]
Should these panels show `Solana Testnet` when doing a Testnet transaction?

https://user-images.githubusercontent.com/40611140/173357925-c549e78d-8798-4354-93dc-71ca48fe1f68.mov

File: components/brave_wallet_ui/utils/solana-instruction-utils.ts (hunk 1 above)
Comment (josheleonard) [reply components/brave_wallet_ui/utils/solana-instruction-utils.ts]: Tests added

File: components/brave_wallet_ui/utils/solana-program-id-utils.ts (hunk 2 above)
Comment (josheleonard) [reply components/brave_wallet_ui/utils/solana-program-id-utils.ts:16]: Tests added

File: components/brave_wallet_ui/components/extension/sign-panel/sign-transaction-panel.tsx [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+  React.useEffect(() => {
+    setSelectedQueueData(signTransactionData?.[0] || undefined)
+  }, [signTransactionData])
+
+  // render
+  return (
+    <StyledWrapper>
+
+      <TopRow>
+        <NetworkText>{'Solana'}</NetworkText>
```
Comment (josheleonard) [inline components/brave_wallet_ui/components/extension/sign-panel/sign-transaction-panel.tsx:192]: Forgot to translate this

Comment (Douglashdaniel) [reply components/brave_wallet_ui/components/extension/sign-panel/sign-transaction-panel.tsx:192]: Updating this to the `Network`!

**Comment by yrliou** [comment]
For non-system instructions which we don't support decode yet, currently it looks like this: (it's a SPL token transfer in this case)
Screen Shot 2022-06-14 at 2 59 46 PM

I think it's better if we can display more details, like showing raw data arrays and account lists.

**Review by Douglashdaniel — APPROVED** [review]
Approval for @josheleonard changes!

**Comment by srirambv** [comment]
Verification passed on 

Brave | 1.42.24 Chromium: 103.0.5060.53 (Official Build) nightly (64-bit) 
-- | --
Revision | `a1711811edd74ff1cf2150f36ffa3b0dae40b17f-refs/branch-heads/5060@{#853}`
OS | Windows 11 Version 21H2 (Build 22000.739)

- Verified test plan from PR
- Verified able to sign and send transactions (single/multiple) 
- Encountered [brave-browser#23663](https://github.com/brave/brave-browser/issues/23663) & [brave-browser#23664](https://github.com/brave/brave-browser/issues/23664)


https://user-images.githubusercontent.com/17010094/175480606-66223ef7-0ab3-4ba7-b341-33cf1a4834ef.mp4

**Pull request merged**